File: node-backend/src/helpers/module_A/pageMetrics/metaExtractor.ts
import type { CheerioAPI } from 'cheerio';
import type { MetaDescriptionData, MetaTagsData, StructuredDataItem } from './types.js';

/**
 * Calculate approximate pixel width of text
 * Based on average character widths for common fonts
 */
function calculatePixelWidth(text: string): number {
    if (!text) return 0;
    
    let width = 0;
    for (const char of text) {
        // Approximate character widths (in pixels)
        if (char === ' ') width += 3;
        else if (/[iIl1\.,;:\-']/.test(char)) width += 4;
        else if (/[fjtJ]/.test(char)) width += 5;
        else if (/[a-z]/.test(char)) width += 6;
        else if (/[A-Z]/.test(char)) width += 7;
        else if (/[wWmM]/.test(char)) width += 9;
        else width += 6; // default
    }
    return Math.round(width);
}

/**
 * Extract meta description from a page
 */
export function extractMetaDescription($: CheerioAPI): MetaDescriptionData {
    const metaDescElement = $('meta[name="description"]');
    const metaDescription = metaDescElement.attr('content')?.trim() || '';
    const metaDescriptionLength = metaDescription.length;
    const hasMissingMetaDescription = metaDescriptionLength === 0;
    
    // Calculate pixel width
    const metaDescriptionPixelWidth = calculatePixelWidth(metaDescription);
    
    return {
        metaDescription: metaDescription || 'No description',
        metaDescriptionLength,
        metaDescriptionPixelWidth,
        hasMissingMetaDescription
    };
}

/**
 * Extract all meta tags from a page
 */
export function extractMetaTags($: CheerioAPI, baseUrl: string): MetaTagsData {
    // Meta keywords
    const metaKeywordsElement = $('meta[name="keywords"]');
    const metaKeywords = metaKeywordsElement.attr('content')?.trim();
    const metaKeywordsLength = metaKeywords?.length;
    
    // Meta robots
    const metaRobotsElement = $('meta[name="robots"]');
    const metaRobots = metaRobotsElement.attr('content')?.trim();
    
    // Canonical URL
    const canonicalElement = $('link[rel="canonical"]');
    let canonicalUrl = canonicalElement.attr('href')?.trim();
    
    // Resolve relative canonical URLs
    if (canonicalUrl && !canonicalUrl.startsWith('http')) {
        try {
            canonicalUrl = new URL(canonicalUrl, baseUrl).toString();
        } catch {
            // Keep original if URL resolution fails
        }
    }
    
    // Viewport
    const viewportElement = $('meta[name="viewport"]');
    const viewport = viewportElement.attr('content')?.trim();
    
    // Meta refresh (HTML-based auto refresh or redirect)
    const metaRefreshElement = $('meta[http-equiv="refresh"]');
    const metaRefresh = metaRefreshElement.attr('content')?.trim();
    
    // Structured data
    const structuredData = extractStructuredData($);
    
    return {
        metaKeywords,
        metaKeywordsLength,
        metaRobots,
        metaRefresh,
        canonicalUrl,
        viewport,
        structuredData
    };
}

/**
 * Extract structured data (JSON-LD, Microdata)
 */
export function extractStructuredData($: CheerioAPI): StructuredDataItem[] {
    const structuredDataItems: StructuredDataItem[] = [];
    
    // Extract JSON-LD
    $('script[type="application/ld+json"]').each((_i: number, el: any) => {
        try {
            const content = $(el).html();
            if (content) {
                const data = JSON.parse(content);
                structuredDataItems.push({
                    type: 'json-ld',
                    schemaType: data['@type'] || data.type,
                    data
                });
            }
        } catch (error) {
            // Invalid JSON or parsing error, skip silently
            // This prevents "require is not defined" errors from breaking the crawl
        }
    });
    
    // TODO: Extract Microdata and RDFa
    // This would require more complex parsing
    
    return structuredDataItems;
}

/**
 * Validate meta description length (SEO best practices)
 */
export function validateMetaDescriptionLength(length: number): {
    isValid: boolean;
    message?: string;
} {
    if (length === 0) {
        return { isValid: false, message: 'Missing meta description' };
    }
    
    if (length < 120) {
        return { isValid: false, message: 'Meta description is too short (< 120 characters)' };
    }
    
    if (length > 160) {
        return { isValid: false, message: 'Meta description is too long (> 160 characters)' };
    }
    
    return { isValid: true };
}

/**
 * Check if canonical URL is valid
 */
export function validateCanonical(canonicalUrl: string | undefined, currentUrl: string): {
    isValid: boolean;
    message?: string;
} {
    if (!canonicalUrl) {
        return { isValid: false, message: 'Missing canonical tag' };
    }
    
    try {
        const canonical = new URL(canonicalUrl);
        const current = new URL(currentUrl);
        
        // Check if canonical points to a different domain
        if (canonical.hostname !== current.hostname) {
            return { 
                isValid: false, 
                message: `Canonical points to different domain: ${canonical.hostname}` 
            };
        }
        
        return { isValid: true };
    } catch {
        return { isValid: false, message: 'Invalid canonical URL format' };
    }
}

File: npy-backend/modules/module_A/ContentAudit/PageMetrics/meta_extractor.py
from bs4 import BeautifulSoup
import re
from urllib.parse import urljoin, urlparse
from modules.module_A.WebsiteCrawler.metrics.pixel_width import calculate_pixel_width as _pw


def _desc_pixel_width(text: str) -> int:
    return _pw(text, font_size=14)

def extract_meta_description(soup: BeautifulSoup) -> dict:
    """
    Extract meta description from a page.
    """
    meta_desc_element = soup.find('meta', attrs={'name': 'description'})
    meta_description = meta_desc_element.get('content', '').strip() if meta_desc_element else ''
    meta_description_length = len(meta_description)
    has_missing_meta_description = meta_description_length == 0
    
    # Calculate pixel width
    meta_description_pixel_width = _desc_pixel_width(meta_description)
    
    return {
        'metaDescription': meta_description or 'No description',
        'metaDescriptionLength': meta_description_length,
        'metaDescriptionPixelWidth': meta_description_pixel_width,
        'hasMissingMetaDescription': has_missing_meta_description
    }

def extract_meta_tags(soup: BeautifulSoup, base_url: str) -> dict:
    """
    Extract all meta tags from a page.
    """
    # Meta keywords
    meta_keywords_element = soup.find('meta', attrs={'name': 'keywords'})
    meta_keywords = meta_keywords_element.get('content', '').strip() if meta_keywords_element else None
    meta_keywords_length = len(meta_keywords) if meta_keywords else 0
    
    # Meta robots
    meta_robots_element = soup.find('meta', attrs={'name': 'robots'})
    meta_robots = meta_robots_element.get('content', '').strip() if meta_robots_element else None
    
    # Canonical URL
    canonical_element = soup.find('link', attrs={'rel': 'canonical'})
    canonical_url = canonical_element.get('href', '').strip() if canonical_element else None
    
    # Resolve relative canonical URLs
    if canonical_url and not canonical_url.startswith('http'):
        try:
            canonical_url = urljoin(base_url, canonical_url)
        except:
            # Keep original if URL resolution fails
            pass
            
    # Viewport
    viewport_element = soup.find('meta', attrs={'name': 'viewport'})
    viewport = viewport_element.get('content', '').strip() if viewport_element else None
    
    # Meta refresh
    meta_refresh_element = soup.find('meta', attrs={'http-equiv': 'refresh'})
    meta_refresh = meta_refresh_element.get('content', '').strip() if meta_refresh_element else None
    
    return {
        'metaKeywords': meta_keywords,
        'metaKeywordsLength': meta_keywords_length,
        'metaRobots': meta_robots,
        'metaRefresh': meta_refresh,
        'canonicalUrl': canonical_url,
        'viewport': viewport
    }

def validate_meta_description_length(length: int) -> dict:
    """
    Validate meta description length (SEO best practices).
    """
    if length == 0:
        return {'isValid': False, 'message': 'Missing meta description'}
    
    if length < 120:
        return {'isValid': False, 'message': 'Meta description is too short (< 120 characters)'}
    
    if length > 160:
        return {'isValid': False, 'message': 'Meta description is too long (> 160 characters)'}
    
    return {'isValid': True}

def validate_canonical(canonical_url: str, current_url: str) -> dict:
    """
    Check if canonical URL is valid.
    """
    if not canonical_url:
        return {'isValid': False, 'message': 'Missing canonical tag'}
    
    try:
        canonical_parsed = urlparse(canonical_url)
        current_parsed = urlparse(current_url)
        
        # Check if canonical points to a different domain
        if canonical_parsed.hostname != current_parsed.hostname:
            return {
                'isValid': False,
                'message': f"Canonical points to different domain: {canonical_parsed.hostname}"
            }
            
        return {'isValid': True}
    except:
        return {'isValid': False, 'message': 'Invalid canonical URL format'}
